File: lib/ftw_realtime/marketplace/project.ex
defmodule FtwRealtime.Marketplace.Project do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id

  schema "projects" do
    field(:name, :string)
    field(:description, :string)

    field(:status, Ecto.Enum,
      values: [:planning, :active, :on_hold, :completed, :cancelled],
      default: :planning
    )

    field(:start_date, :date)
    field(:end_date, :date)
    field(:budget, :integer, default: 0)
    field(:spent, :integer, default: 0)

    belongs_to(:contractor, FtwRealtime.Marketplace.User)
    belongs_to(:homeowner, FtwRealtime.Marketplace.User)
    belongs_to(:job, FtwRealtime.Marketplace.Job)

    timestamps(type: :utc_datetime)
  end

  def changeset(project, attrs) do
    project
    |> cast(attrs, [
      :name,
      :description,
      :status,
      :start_date,
      :end_date,
      :budget,
      :spent,
      :contractor_id,
      :homeowner_id,
      :job_id
    ])
    |> validate_required([:name, :contractor_id, :homeowner_id])
    |> foreign_key_constraint(:contractor_id)
    |> foreign_key_constraint(:homeowner_id)
    |> foreign_key_constraint(:job_id)
  end
end

File: src/main/kotlin/com/strata/ftw/domain/entity/Project.kt
package com.strata.ftw.domain.entity

import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.Instant
import java.time.LocalDate
import java.util.UUID

@Entity
@Table(name = "projects")
class Project(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(columnDefinition = "uuid")
    var id: UUID? = null,

    @Column(nullable = false)
    var name: String = "",

    var description: String? = null,

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    var status: ProjectStatus = ProjectStatus.planning,

    @Column(name = "start_date")
    var startDate: LocalDate? = null,

    @Column(name = "end_date")
    var endDate: LocalDate? = null,

    var budget: Int = 0,
    var spent: Int = 0,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "contractor_id")
    var contractor: User? = null,

    @Column(name = "contractor_id", insertable = false, updatable = false)
    var contractorId: UUID? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "homeowner_id")
    var homeowner: User? = null,

    @Column(name = "homeowner_id", insertable = false, updatable = false)
    var homeownerId: UUID? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "job_id")
    var job: Job? = null,

    @Column(name = "job_id", insertable = false, updatable = false)
    var jobId: UUID? = null,

    var category: String? = null,

    @OneToMany(mappedBy = "project", fetch = FetchType.LAZY)
    @OrderBy("sortOrder ASC")
    var milestones: MutableList<Milestone> = mutableListOf(),

    @OneToMany(mappedBy = "project", fetch = FetchType.LAZY)
    @OrderBy("date DESC")
    var expenses: MutableList<Expense> = mutableListOf(),

    @CreationTimestamp
    @Column(name = "inserted_at", updatable = false)
    var insertedAt: Instant? = null,

    @UpdateTimestamp
    @Column(name = "updated_at")
    var updatedAt: Instant? = null
)

enum class ProjectStatus {
    planning, active, on_hold, completed, cancelled
}
